src/tokio/caller.rs

```diff
@@ -0,0 +1,135 @@
+
+        let mut state = SrtConnectionState {
+            // this needs to be fused here so it actually doesn't get polled after
+            // completion
+            // the streams don't need this as we construct a new futures each time[
+            close_rx: close_rx.fuse(),
+            srt_send_rx,
+            socket,
+            sender,
+        };
+
+        let task = tokio::spawn(async move {
+            state.run().await;
+        });
+
+        Ok(SrtCaller {
+            close_tx: Some(close_tx),
+            srt_send_tx: srt_send_tx,
+            _task: task,
+        })
+    }
+
+    pub async fn send(&mut self, data: Bytes) -> Result<(), mpsc::SendError> {
+        self.srt_send_tx.send((data, Instant::now())).await
+    }
+
+    pub async fn shutdown(&mut self) -> Result<(), Error> {
+        self.close_tx = None;
+        Ok(())
+    }
+}
+
+struct SrtConnectionState {
+    close_rx: Fuse<oneshot::Receiver<()>>,
+    srt_send_rx: mpsc::Receiver<(Bytes, Instant)>,
+    socket: UdpFramed<PacketCodec>,
+    sender: Sender,
+}
+
+impl SrtConnectionState {
+    async fn send_control(&mut self, packet: ControlPacket, to: SocketAddr) {
+        let result = self.socket.send((Packet::Control(packet), to)).await;
+        let _ = result.map_err(|error| eprintln!("{:?}", error));
+    }
+
+    async fn send_data(&mut self, packet: DataPacket, to: SocketAddr) {
+        let result = self.socket.send((Packet::Data(packet), to)).await;
+        let _ = result.map_err(|error| eprintln!("{:?}", error));
+    }
+
+    async fn wait(&mut self) {
+        // TODO: rethink magic number
+        self.wait_until(Instant::now() + Duration::from_millis(10))
+            .await
+    }
+
+    async fn wait_until(&mut self, t: Instant) {
+        // TODO: get rid of the panics with legit error handling, e.g. an error channel on SrtCaller
+        let result = select! {
+            _ = &mut self.close_rx => self.sender.handle_close(Instant::now()),
+            _ = delay_until(t.into()).fuse() => self.sender.handle_timer(Instant::now()),
+            packet = crate::util::get_packet(&mut self.socket).fuse() => self.sender.handle_packet(packet.unwrap(), Instant::now()),
+            data = self.srt_send_rx.next().fuse() => self.sender.handle_send_rx(data.unwrap(), Instant::now()),
+        };
+        let _ = result.map_err(|error| eprintln!("{:?}", error));
+    }
+
+    pub async fn run(&mut self) {
+        use SenderAlgorithmAction::*;
+        loop {
+            let (current_step, next_action) = self.sender.next_algorithm_action(|| Instant::now());
+            debug!("{:?}: {:?}", current_step, next_action);
+            match next_action {
+                Continue => {}
+                SendControl((packet, to)) => self.send_control(packet, to).await,
+                SendData((packet, to)) => self.send_data(packet, to).await,
+                WaitUnitlAck => self.wait().await,
+                WaitUntilData => self.wait().await,
```

Comment:
Combine these cases

Reply:
good point, as I was refactoring, I was kind of holding out that there would actually be a difference, but nope..